azurelinuxagent/common/cgroups.py

```diff
@@ -0,0 +1,689 @@
+    without much state worth preserving; it's not unreasonable to create one just when you need it.
+    """
+    # whether cgroup support is enabled
+    _enabled = True
+    _hierarchies = CGroupsTelemetry.metrics_hierarchies()
+    _use_systemd = None     # Tri-state: None (i.e. "unknown"), True, False
+    _osutil = get_osutil()
+
+    @staticmethod
+    def _construct_custom_path_for_hierarchy(hierarchy, cgroup_name):
+        return os.path.join(BASE_CGROUPS, hierarchy, AGENT_NAME, cgroup_name).rstrip(os.path.sep)
+
+    @staticmethod
+    def _construct_systemd_path_for_hierarchy(hierarchy, cgroup_name):
+        return os.path.join(BASE_CGROUPS, hierarchy, 'system.slice', cgroup_name).rstrip(os.path.sep)
+
+    @staticmethod
+    def for_extension(name):
+        return CGroups(name, CGroups._construct_custom_path_for_hierarchy)
+
+    @staticmethod
+    def for_systemd_service(name):
+        return CGroups(name.lower(), CGroups._construct_systemd_path_for_hierarchy)
+
+    def __init__(self, name, path_maker):
+        """
+        Construct CGroups object. Create appropriately-named directory for each hierarchy of interest.
+
+        :param str name: Name for the cgroup (usually the full name of the extension)
+        :param path_maker: Function which constructs the root path for a given hierarchy where this cgroup lives
+        """
+        if name == "":
+            self.name = "Agents+Extensions"
+            self.is_wrapper_cgroup = True
+        else:
+            self.name = name
+            self.is_wrapper_cgroup = False
+
+        self.cgroups = {}
+
+        if not self.enabled():
+            return
+
+        system_hierarchies = os.listdir(BASE_CGROUPS)
+        for hierarchy in CGroups._hierarchies:
+            if hierarchy not in system_hierarchies:
+                raise CGroupsException("Hierarchy {0} is not mounted".format(hierarchy))
+
+            cgroup_name = "" if self.is_wrapper_cgroup else self.name
+            cgroup_path = path_maker(hierarchy, cgroup_name)
+            if not os.path.isdir(cgroup_path):
+                logger.info("Creating cgroup directory {0}".format(cgroup_path))
+                CGroups._try_mkdir(cgroup_path)
+            self.cgroups[hierarchy] = cgroup_path
+
+    @staticmethod
+    def is_systemd_manager():
+        """
+        Determine if systemd is managing system services. Many extensions are structured as a set of services,
+        including the agent itself; systemd expects those services to remain in the cgroups in which it placed them.
+        If this process (presumed to be the agent) is in a cgroup that looks like one created by systemd, we can
+        assume systemd is in use.
+
+        :return: True if systemd is managing system services
+        :rtype: Bool
+        """
+        if CGroups._use_systemd is None:
+            hierarchy = METRIC_HIERARCHIES[0]
```

Comment:
I know that [0] is cpu, but it isn't a very constant.

Reply:
I don't care which hierarchy it is; it just has to be one of the hierarchies from which we're selecting metrics. It could be "memory" for all I care, and the code will work.